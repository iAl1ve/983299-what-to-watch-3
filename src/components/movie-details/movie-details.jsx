import React, {memo} from "react";
import {func, object} from "prop-types";
import {Route, Switch, Link, withRouter} from "react-router-dom";
import MovieList from "../movie-list/movie-list.jsx";
import Tabs from "../tabs/tabs.jsx";
import {FilmsType, FilmType} from "../../types";
import withActiveItem from '../../hocs/with-active-item/with-active-item.jsx';

const MovieListWithActiveItem = withActiveItem(MovieList);

const MovieDetails = ({match, cardData: {name, img} = {}, onOpenCard, filteredFilms}) => {
  const {path, url} = match;

  return (
    <>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={img} alt={name} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <header className="page-header movie-card__head">
            <div className="logo">
              <a href="main.html" className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </a>
            </div>

            <div className="user-block">
              <div className="user-block__avatar">
                <img
                  src="img/avatar.jpg"
                  alt="User avatar"
                  width="63"
                  height="63"
                />
              </div>
            </div>
          </header>

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">Drama</span>
                <span className="movie-card__year">2014</span>
              </p>

              <div className="movie-card__buttons">
                <button
                  className="btn btn--play movie-card__button"
                  type="button"
                >
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button
                  className="btn btn--list movie-card__button"
                  type="button"
                >
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
                <a href="add-review.html" className="btn movie-card__button">
                  Add review
                </a>
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img
                src="img/the-grand-budapest-hotel-poster.jpg"
                alt="The Grand Budapest Hotel poster"
                width="218"
                height="327"
              />
            </div>

            <div className="movie-card__desc">
              <nav className="movie-nav movie-card__nav">
                <ul className="movie-nav__list">
                  <li className="movie-nav__item">
                    <Link to={`${url}`} className="movie-nav__link">
                      Overview
                    </Link>
                  </li>
                  <li className="movie-nav__item">
                    <Link to={`${url}/details`} className="movie-nav__link">
                      Details
                    </Link>
                  </li>
                  <li className="movie-nav__item">
                    <Link to={`${url}/reviews`} className="movie-nav__link">
                      Reviews
                    </Link>
                  </li>
                </ul>
              </nav>

              <div>
                <Switch>
                  <Route exact path={`${path}`}>
                    <Tabs />
                  </Route>
                  <Route path={`${path}/:tab`}>
                    <Tabs />
                  </Route>
                </Switch>
              </div>
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <MovieListWithActiveItem onOpenCard={onOpenCard} filteredFilms={filteredFilms}/>
        </section>
      </div>
    </>
  );
};
MovieDetails.propTypes = {
  match: object,
  cardData: FilmType,
  films: FilmsType,
  onOpenCard: func,
  filteredFilms: FilmsType
};
export {MovieDetails};
export default withRouter(memo(MovieDetails));
