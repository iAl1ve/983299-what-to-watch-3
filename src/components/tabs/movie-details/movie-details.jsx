import React from 'react';
import PropTypes from 'prop-types';

const getStarsListHtml = (stars) => {
  let newStarsList = [];
  stars.map((star, i) => [`${star}, `, <br key={`${i}-${star}`}></br>]).forEach((arr) => {
    newStarsList = newStarsList.concat(arr);
  });
  newStarsList.pop();
  newStarsList[newStarsList.length - 1] = newStarsList[newStarsList.length - 1].slice(0, -2);
  return newStarsList;
};

const convertFilmDuration = (time) => {
  const hours = Math.floor(time / 60);
  const minutes = time % 60;
  return `${hours}h ${minutes}m`;
};

const MovieDetails = ({film}) => {
  const {director, starring, filmDuration, genre, releaseYear} = film;
  let newStarsList = [];
  starring.map((star, i) => [`${star}, `, <br key={`${i}-${star}`}></br>]).forEach((arr) => {
    newStarsList = newStarsList.concat(arr);
  });
  newStarsList.pop();
  newStarsList[newStarsList.length - 1] = newStarsList[newStarsList.length - 1].slice(0, -2);
  return (
    <div className="movie-card__text movie-card__row">
      <div className="movie-card__text-col">
        <p className="movie-card__details-item">
          <strong className="movie-card__details-name">Director</strong>
          <span className="movie-card__details-value">{director}</span>
        </p>
        <p className="movie-card__details-item">
          <strong className="movie-card__details-name">Starring</strong>
          <span className="movie-card__details-value">
            {getStarsListHtml(starring)}
          </span>
        </p>
      </div>

      <div className="movie-card__text-col">
        <p className="movie-card__details-item">
          <strong className="movie-card__details-name">Run Time</strong>
          <span className="movie-card__details-value">{convertFilmDuration(filmDuration)}</span>
        </p>
        <p className="movie-card__details-item">
          <strong className="movie-card__details-name">Genre</strong>
          <span className="movie-card__details-value">{genre}</span>
        </p>
        <p className="movie-card__details-item">
          <strong className="movie-card__details-name">Released</strong>
          <span className="movie-card__details-value">{releaseYear}</span>
        </p>
      </div>
    </div>
  );
};

MovieDetails.propTypes = {
  film: PropTypes.shape({
    title: PropTypes.string,
    genre: PropTypes.string,
    releaseYear: PropTypes.number,
    imgSrc: PropTypes.string,
    bgSrc: PropTypes.string,
    posterSrc: PropTypes.string,
    ratingScore: PropTypes.number,
    ratingCount: PropTypes.number,
    description: PropTypes.arrayOf(PropTypes.string),
    director: PropTypes.string,
    starring: PropTypes.arrayOf(PropTypes.string),
    id: PropTypes.number,
    filmDuration: PropTypes.number,
    reviews: PropTypes.array,
  })
};

export default MovieDetails;
