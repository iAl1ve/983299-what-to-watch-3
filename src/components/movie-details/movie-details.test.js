import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {MovieDetails} from './movie-details';
import {MemoryRouter} from 'react-router';

const mockStore = configureStore([]);

const cardData = {
  name: `test`,
  img: `img/test.jpg`
};

const films = [
  {name: `test1`, img: `test1.img`, genre: `test1Genre`},
  {name: `test2`, img: `test2.img`, genre: `test2Genre`},
  {name: `test3`, img: `test3.img`, genre: `test3Genre`},
];

const match = {
  path: `testPath`,
  url: `testURL`
};

it(`MovieDetails renders correctly`, () => {
  const store = mockStore({
    genres: [`genre1`, `genre2`],
    genreFilter: `All genres`,
    films,
    filteredFilms: films,
  });

  const tree = renderer
    .create(
        <MemoryRouter initialEntries={[`/test`]} >
          <Provider store={store}>
            <MovieDetails cardData={cardData} films={films} match={match} filteredFilms={films}/>
          </Provider>
        </MemoryRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
