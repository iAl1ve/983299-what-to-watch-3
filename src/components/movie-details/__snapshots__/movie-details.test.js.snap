// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="test"
          src="img/test.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="main.html"
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="img/avatar.jpg"
              width="63"
            />
          </div>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            test
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="#d9cd8d"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5v14l11-7z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="add-review.html"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/testURL"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/testURL/details"
                  onClick={[Function]}
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/testURL/reviews"
                  onClick={[Function]}
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div />
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="test1"
              height="175"
              src="test1.img"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <a
              className="small-movie-card__link"
              href="/dev-component"
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "display": "flex",
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              test1
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="test2"
              height="175"
              src="test2.img"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <a
              className="small-movie-card__link"
              href="/dev-component"
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "display": "flex",
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              test2
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="test3"
              height="175"
              src="test3.img"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <a
              className="small-movie-card__link"
              href="/dev-component"
              onClick={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "display": "flex",
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              test3
            </a>
          </h3>
        </article>
      </div>
    </section>
  </div>,
]
`;
